### PR TITLE
Fix/batch/filenames

### LINKED ISSUE #153: Batch can't access online files
Somehow during one of the refactoring batch is no longer able to open online files. Here's the error I [get on this colab](https://colab.research.google.com/drive/1VOlWL_0Z6IS6_A-5lwSB5U-u-4yvP3aR#scrollTo=UrpQ3eyB7oeo) (that used to work). I imagine we're not using the right function in `endaq.ide` @StokesMIDE @pscheidler 

~~~python
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-11-9abaa8ef6763> in <module>()
      7     #.add_pvss(init_freq=1,bins_per_octave=0.001)
      8     .add_metrics()
----> 9     .aggregate_data(files)
     10 )
     11 file_data = calc_output.dataframes

3 frames
/usr/local/lib/python3.7/dist-packages/endaq/batch/core.py in aggregate_data(self, filenames)
    519 
    520         series_lists = zip(
--> 521             *(self._get_data(filename).values() for filename in filenames)
    522         )
    523 

/usr/local/lib/python3.7/dist-packages/endaq/batch/core.py in <genexpr>(.0)
    519 
    520         series_lists = zip(
--> 521             *(self._get_data(filename).values() for filename in filenames)
    522         )
    523 

/usr/local/lib/python3.7/dist-packages/endaq/batch/core.py in _get_data(self, filename)
    467 
    468         data = {}
--> 469         with endaq.ide.get_doc(filename) as ds:
    470             ch_data_cache = endaq.batch.analyzer.CalcCache.from_ide(
    471                 ds,

/usr/local/lib/python3.7/dist-packages/endaq/ide/files.py in get_doc(name, filename, url, parsed, start, end, localfile, params, headers, cookies, **kwargs)
    183     if filename:
    184         filename = os.path.abspath(os.path.expanduser(filename))
--> 185         stream = open(filename, 'rb')
    186 
    187     if stream:

FileNotFoundError: [Errno 2] No such file or directory: '/content/https:/info.endaq.com/hubfs/data/DAQ11023_002778.ide'
~~~

### PR DESCRIPTION
**Pull request by CrepeGoat**
resolves #153 

attn @shanlyMIDE

### CODE REVIEW

File: endaq/batch/core.py [hunk 1]
```diff
@@ -505,27 +505,45 @@ def _get_data(self, filename):
...
-        file_display_names = [
-            os.path.relpath(name, start=root_path) for name in filenames
+        if len(filenames) == 0:
+            return None
+
+        http_files, local_files = [], []
+        for file in filenames:
+            (
+                http_files
+                if file.startswith("http://") or file.startswith("https://")
```
Comment (CrepeGoat) [inline endaq/batch/core.py:521]: @StokesMIDE is this an acceptable check for whether a file is a network or local file?

Comment (StokesMIDE) [reply endaq/batch/core.py:521]: It's probably fine, although the filename *could* be a `file:` URL, which would need parsing... you might want to consider something like what I did in `get_doc()`: https://github.com/MideTechnology/endaq-python/blob/d2196257c96ce6e0297d01de33c65dbd1c8340ea/endaq/ide/files.py#L157

Comment (StokesMIDE) [reply endaq/batch/core.py:521]: File URLs can be generated if a filename is copied to/from an email or something, so it isn't entirely unlikely.

**Comment by shanlyMIDE** [comment]
Looks good, I tested it here in this colab: https://colab.research.google.com/drive/1VOlWL_0Z6IS6_A-5lwSB5U-u-4yvP3aR#scrollTo=7nw2iCh89dC0

File: endaq/batch/core.py (hunk 1 above)
Comment (CrepeGoat) [reply endaq/batch/core.py]: yeah I considered it, but I wasn't sure how deep I wanted to get into the issue. how about a7c3b456aaf8ee8a21ea4be84a8b7c84b060264e?

File: endaq/ide/files.py [hunk 2]
```diff
@@ -20,7 +25,49 @@
...
+
+# ============================================================================
+#
+# ============================================================================
+
+
+def normalized_path(
+    filepath: str, mode: typing.Literal[None, "local", "url"] = None
+) -> Union[
+    Tuple[str, typing.Literal["local"]], Tuple[ParseResult, typing.Literal["url"]]
```
Comment (CFlaniganMide) [inline endaq/ide/files.py:38]: I'm really gonna nitpick here, sorry.  Can we separate longer `Union` statements onto separate lines, such that each option for the union is separate from the rest?

**Review by CFlaniganMide — APPROVED** [review]
Leaving my approval since all I have is a formatting quibble and David's got more informed opinions on the meat of this PR.

File: endaq/ide/files.py (hunk 2 above)
Comment (CrepeGoat) [reply endaq/ide/files.py]: you meant like ab156e1434c1a2bd0b73cbf2f356d657fc2d6b36, right?

File: endaq/batch/core.py (hunk 1 above)
Comment (CrepeGoat) [reply endaq/batch/core.py]: @StokesMIDE I got an approval, but is there anything you wanted to discuss here?

**Review by StokesMIDE — APPROVED** [review]
The implementation of `normalized_path()` seems a bit complicated when it looks like it was implemented mainly to create a common path. But maybe it's just because it's a change to my code and not how I would have done it.

**Comment by CrepeGoat** [comment]
> The implementation of `normalized_path()` seems a bit complicated when it looks like it was implemented mainly to create a common path. But maybe it's just because it's a change to my code and not how I would have done it.

yeah tbh I'm not in love with it either, but I'm not sure how best to change it. any ideas are welcome for sure! but maybe for now we just keep it?

File: endaq/batch/core.py [hunk 3]
```diff
@@ -505,27 +505,45 @@ def _get_data(self, filename):
 
         return data
 
-    def aggregate_data(self, filenames) -> OutputStruct:
+    def aggregate_data(self, filenames) -> Optional[OutputStruct]:
```
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply endaq/batch/core.py:508]: suggestion noted, thanks. personally I think this is a style choice, and I can't see a good reason not to use a std-provided annotation

[the remark this answers is not in the document]
Comment (StokesMIDE) [reply endaq/batch/core.py:508]: `Optional` *is* standard, but it's intended for different use. It's a linguistic thing... return values aren't something that can be optional; it's not under user control the way parameters are, and functions will always return something.

**Comment by StokesMIDE** [comment]
> yeah tbh I'm not in love with it either, but I'm not sure how best to change it. any ideas are welcome for sure! but maybe for now we just keep it?

My solution would require some non-trivial refactoring of `aggregate_data()`, but really, this is plumbing, not user-facing. We can potentially revise things later.

**Pull request merged**